Added initial set of tests for properties of CSS Borders 4

SebastianZ opened the pull request:
This adds an initial set of tests related to the CSS properties introduced in [CSS Borders 4](https://drafts.csswg.org/css-borders-4/).

Sebastian

SebastianZ committed:
Added initial set of tests for properties of CSS Borders 4

SebastianZ commented:
@fantasai @LeaVerou @Loirooriol That's the first time I am adding web platform tests. Maybe you could review them?

(As I am not a contributor to this project yet, I am not allowed to add reviewers directly to this PR. That's why I wrote this comment instead.)

Sebastian

TalbotG commented:
@SebastianZ Sebastian,

I can do some review for you :)

1- 
The lint integration check failed. In other words, this is an important task and an important integration check to pass. You must pass such test before "earning" the right for a review from reviewers :)

As far as I can see, the lint integration check failed because 

"
Test-file line starts with one or more tab characters (INDENT TABS)
"

So, you must remove those "nasty" tab characters at the beginning of lines.

"
line 951: INFO:lint:There were 842 errors (INDENT TABS: 842)
"
coming from the Lint log

"
INDENT TABS

    Test-file line starts with one or more tab characters
    To fix:

    use spaces to replace any tab characters at 
    beginning of lines
"
coming from
http://web-platform-tests.org/writing-tests/lint-tool.html

"
Use spaces rather than tabs for indentation
"
coming from
http://web-platform-tests.org/writing-tests/general-guidelines.html#style-rules

2- 
`<meta charset="utf-8" />`
There is no need for the Null End Tag everywhere in your tests. It is not a mistake; it is just not needed, not necessary.
`<meta charset="utf-8">`
is good.

3-
`<p>Test passes if there are four black boxes with rounded corners and red is not visible on the page.</p>`

I recommend to embold no red, to use `<strong>no red</strong>` so that this pass conditions stands out. 

"visible on the page" is pleonastic, not useful, self-evident.
That's item 12 in my list in
https://lists.w3.org/Archives/Public/public-css-testsuite/2018Jun/0000.html
"
Why remove such words? Because they are useless, empty-meaning, 
pleonastic or self-evident or irrelevant words; they can be safely 
removed without affecting understanding of pass/fail conditions. 
Anything that shortens the pass/fail conditions sentence of a test, 
without compromising its understanding for a tester, helps.
"

In HTML5, the closing tag `</p>` can be removed too. I have been personally requested to remove all optional closing tags from my tests.
Even `<head>` , `<body>` and `<html>` are optional and so I removed them all from my tests: I have been requested to do so.

4-
New line missing at end of file of many of your tests. You need to add an empty line at end of file of your tests. If it is missing, then you will see an icon of a red circle with minus sign inside of it when the reviewer views your tests in the Commits tab .

In all fairness, this should be in a documentation file...

5-
For title text, I recommend to reuse the important description words of the spec like this eg:
  `<title>CSS Borders and Box Decorations Test: Logical 'border-radius-*' shorthands set the two radii of one side.</title>`
instead of
  `<title>CSS Test: Logical 'border-radius-*' shorthands set the two radii of one side.</title>`

6-
Meta test text assert. This is optional. If the test is complex or involving some particular code scenario, then you may need to write some explanation with the goal of helping the reviewer and the eventual testers. I recommend to start with the following

`<meta name="assert" content="This test checks that ...">`

In your corner-shape-angle.html , I would say that your meta test text assert is not needed or not decisive. What you did is not a mistake.

"
The assertion should not be:

    A copy of the title text

    A copy of the test verification instructions

    A duplicate of another assertion in the test suite

    A line or reference from the CSS specification 
    unless that line is a complete assertion when 
    taken out of context.

The test assertion is optional, but is highly recommended. It helps the reviewer understand the goal of the test so that he or she can make sure it is being tested correctly. Also, in case a problem is found with the test later, the testing method (e.g. using color to determine pass/fail) can be changed (e.g. to using background-color) while preserving the intent of the test (e.g. testing support for ID selectors).

Examples of good test assertions:

    “This test checks that a background image with 
    no intrinsic size covers the entire padding box.”

    “This test checks that ‘word-spacing’ affects each 
    space (U+0020) and non-breaking space (U+00A0).”

    “This test checks that if ‘top’ and ‘bottom’ offsets are 
    specified on an absolutely-positioned replaced element, 
    then any remaining space is split amongst the ‘auto’ 
    vertical margins.”

    “This test checks that ‘text-indent’ affects only the 
    first line of a block container if that line is also the 
    first formatted line of an element.”

"
coming from
http://web-platform-tests.org/writing-tests/css-metadata.html#test-assertions


7-
Filenaming.
corner-shape/corner-shape-round.html
It is best to end with a number, like, eg, 001 in filenames. So that if you or someone else want to write another test involving a rounded corner, then such person only has to continue from the last number in the filename. 

"
A common format is test-topic-001.html, where test-topic is a short identifier that describes the test. It should avoid conjunctions, articles, and prepositions as it should be as concise as possible. The integer that follows is normally just increased incrementally, and padded to three digits. (If you’d end up with more than 999 tests, your test-topic is probably too broad!)
"
coming from
http://web-platform-tests.org/writing-tests/general-guidelines.html#file-paths-and-names

Same issue with border-clip-valid.html border-clip-invalid.html and maybe other tests. What you did is not necessarly a mistake but it is not ideal.

**Conclusion**
the most important things for you right now is to correct the 842 tab characters (INDENT TABS) and add an empty line at EOF of several of your tests.

TalbotG commented:
8-
Some (several?) of your tests exceed the 600px height window viewport constraint.
"
The test renders within a 800x600 viewport
"
coming from
http://web-platform-tests.org/reviewing-tests/checklist.html#visual-tests-only

Eg
```
  <div id="reference1"></div>
  <div id="reference2"></div>
  <div id="reference3"></div>
  <div id="reference4"></div>
```
in border-radius-side-shorthands-001-ref.html
where each div is 200px tall with a vertical margin of 10px ~= 986px.

Only the first 600px of height of screen shots are compared.

The documentation says the viewport must be (maximum) 800px wide and 600px tall but even that is not clear. Elsewhere it is written 600px by 600px.

Solution: smaller boxes or 2 rows of 2 boxes side-by-side or even split tests in 2 sub-tests

9-
In 
border-radius-side-shorthands-001-ref.html
the borders of the 4 boxes should be black according to pass-fail condition sentence but their borders are definitely red because the code says so!

In the test
border-radius-side-shorthands-001.html
it is not possible for the test to display red. So
the pass-fail condition sentence is not right, is not consequent.

SebastianZ committed:
Fixed formatting issues in css-borders-4 tests

SebastianZ commented:
Thank you @TalbotG for your review! I'll try to fix and improve all the things you mentioned over the next few days. As a start, I've addressed the linting issues.

Sebastian

TalbotG commented:
@SebastianZ

At line 20 of the file 
box-shadow-offset-computed.html
I see:
`test_computed_value("box-shadow-offset", "1F0px 20px, 30px 40px");`
The "**F**" should not be there.

TalbotG commented:
In
corner-shape-mixed-values-001.html
you are testing
'corner-shape: round angle'
which is 2 values while the text assert
says 
"Three different values are supported for 'corner-shape'"

- - -

You have tested
'corner-shape: angle'
'corner-shape: round angle'
'corner-shape: round angle round'
'corner-shape: round angle round angle'
'corner-shape: round'

Do you intend to also test other mixed values like
'corner-shape: angle round'
'corner-shape: angle round angle'
'corner-shape: angle round angle round'
and
'corner-shape: round angle angle angle'
'corner-shape: round angle angle round'
'corner-shape: round round angle angle'
etc.
?

SebastianZ committed:
Corrected remaining issues in css-borders-4 tests

SebastianZ committed:
Added tests for all combinations of 'corner-shape' values

SebastianZ commented:
@TalbotG I think I've addressed all your feedback. Please let me know if there's anything missing!

> 7-
> Filenaming.
> ...
> Same issue with border-clip-valid.html border-clip-invalid.html and maybe other tests. What you did is not necessarly a mistake but it is not ideal.

That's the only thing I didn't change as basically all other CSS spec. tests have this *-valid.html, *-invalid.html, *-computed.html naming convention in their parsing/ subfolder, even the newest ones like css-view-transitions.

Sebastian

PS: I really don't like leaving out tags like `<html>`, `</p>` and prefer a more XML-like syntax, though as most of the tests seem to do that I followed your advice and adjusted the structure accordingly.

TalbotG commented:
Sebastian,

10-
You can safely remove the words "on the page" from the pass-fail conditions sentence everywhere. Since 95% (or so) of all tests starts with a sentence stating the pass and fail conditions, then 
"in this page",
"on this page",
"on the screen",
"below", 
"below this line", 
"after this line",
"after this sentence", 
"below this paragraph",
"under this paragraph", 
"after this",
"following",
"which follows",
etc.. 
is evident, is pleonastic, is not useful.

11-
Only the 2 border-radius-side-shorthands tests needed to have shorter `div` because it seemed that only those 2 were creating taller than 600px document box. What you did (reducing all the divs from 200px to 100px everywhere) is not a mistake.

12-
For the test
corner-shape-round-001.html
I suspect it is possible to create a reference for it... Do you agree?

13-
In box-shadow-blur-computed.html tests
you can safely remove
line 6: `<meta name="assert" content="Computed value of 'box-shadow-blur' is correct">`
because such test assert text is not helpful, not meaningful, not worthy, not significant.

Same thing with
`<meta name="assert" content="Computed value of 'box-shadow-color' is correct">`
in box-shadow-color-computed.html

Also in
box-shadow-offset-computed.html , box-shadow-position-computed.html , box-shadow-spread-computed.html and corner-shape-computed.html

TalbotG commented:
14-
Reference files should go in (move to) a /reference folder: this is best. What you did is not a mistake. But it is better to have such folder separation.

TalbotG commented:
> > Same issue with border-clip-valid.html border-clip-invalid.html and maybe other tests. What you did is not necessarly a mistake but it is not ideal.
> 
> That's the only thing I didn't change as basically all other CSS spec. tests have this *-valid.html, *-invalid.html, *-computed.html naming convention in their parsing/ subfolder, even the newest ones like css-view-transitions.

Then fine. Leave these tests' filenames as they are. No problem.

> 
> Sebastian
> 
> PS: I really don't like leaving out tags like `<html>`, `</p>` and prefer a more XML-like syntax

I am like you. I prefer to edit `<html>`, `<body>`,  `</p>` and all optional closing tags (`</tr>`, `</td>`, etc.) but I was told and requested to make that change. If you visit my website, you will see that I use HTML4.01 strict almost everywhere.

SebastianZ committed:
Incorporated further feedback on css-borders-4 tests

SebastianZ commented:
Made the requested changes.

> 11- Only the 2 border-radius-side-shorthands tests needed to have shorter `div` because it seemed that only those 2 were creating taller than 600px document box. What you did (reducing all the divs from 200px to 100px everywhere) is not a mistake.

Sure, though I went with your general rule for making squares 100px wide to emphasize the borders.

> 12- For the test corner-shape-round-001.html I suspect it is possible to create a reference for it... Do you agree?

Yes, done that not only for that test but als for the others with completely rounded corners.

> 13- In box-shadow-blur-computed.html tests you can safely remove line 6: `<meta name="assert" content="Computed value of 'box-shadow-blur' is correct">` because such test assert text is not helpful, not meaningful, not worthy, not significant.
> ...

Ok, removed the assertions from those tests and improved the sentences in the other tests.

Sebastian

TalbotG commented:
One last modification...

15- Green and red colors should always be linked, should be hand-in-hand, should be mirrored. If you use green color in a test, then red color should be involved in the test. If a test's pass-fail conditions sentence state "(...) no red", then it must use the green color to indicate success. This is an established convention, more or less a standard now when doing CSS tests. It has been the case since the beginning of reftests over at W3C.

"
Red Means Failure
    Don't use the color red other than to indicate a failure. (Exception: testing for support of red) Since green-with-no-red is often used to indicate success, **it's best to also avoid green unless using the presence of red to indicate failures.**
"
coming from
https://wiki.csswg.org/test/format#design-requirements

"
In general, using colors in a consistent manner is recommend. Specifically, the following convention has been developed: 

Red
    Any red indicates failure. 
Green
    In the absence of any red, green indicates success.
"
coming from 2003's CSS2.1 Test Case Authoring Guidelines document
https://www.w3.org/Style/CSS/Test/guidelines.html#color

So the 2 border-radius-side-shorthands tests and their related references should be updated accordingly. Just replace where needed "black" with "green".

SebastianZ commented:
> 15- Green and red colors should always be linked
> ...
> So the 2 border-radius-side-shorthands tests and their related references should be updated accordingly. Just replace where needed "black" with "green".

Makes sense. Done.

Sebastian

TalbotG commented:
@SebastianZ Sebastian,

"there are four black squares"
must become
"there are four green squares"
in the 2 border-radius-side-shorthands tests 
and in their related references.

I expect to be away and 
less available in the next 3 days.

SebastianZ commented:
> @SebastianZ Sebastian,
> 
> "there are four black squares" must become "there are four green squares" in the 2 border-radius-side-shorthands tests and in their related references.

Oops! But it's ok for the black-green colorblind people.😆 Fixed it now.

> I expect to be less available in the next 3 days.

No worries. There's no hurry.

Sebastian

TalbotG commented:
@SebastianZ Sebastian,

I have not forgotten you. I am in the middle of computer migration and personal matter that require my time and undivided attention. I will make a final review of your tests - now that all formatting issues that I could find have been all covered and corrected by you - later this week: this upcoming Thursday February 1st or Friday February 2nd.

TalbotG commented:
16-
In border-radius-side-shorthands-001.html
I see
line 77:  `border-radius-right: 1 4em / 5em 1em;`
The "1" should be "2em" .

17-
In all of corner-shape-four-values tests (except 001):
if there is no reference file for those tests, 
then you must append "-manual" to
such tests: "a -manual suffix after the main 
filename but before the extension".

More info:
http://web-platform-tests.org/writing-tests/manual.html

There may be a way to create reference files
for each and all corner-shape-* tests but, right now, 
I do not know.

18-
In corner-shape-four-values-001
meta name="assert" content="This test checks that 'round round round round' is supported for 'corner-shape'."
but, in reality, the test is currently passed by Firefox and Chromium because of a fallback feature: if a property (and/or a property value) is not implemented, then it is ignored. This is why the test apparently and seemingly passes right now in Firefox and Chromium (and also - I have not checked - most likely pass in Safari) but corner-shape is not implemented. So, this particular test assert text is not true, is not correct. I think now you need to remove that test after all...

19-
In tests and/or in references, if you can combine selectors,
then this is preferable.
Eg
In border-radius-side-shorthands-001-ref.html
in line 20 to 46, the code can become:
```

#reference-shape-1 , #reference-shape-3 {
  border-top-left-radius: 40px;
  border-top-right-radius: 40px;
  border-bottom-right-radius: 40px;
  border-bottom-left-radius: 40px;
}

#reference-shape-2 , #reference-shape-4 {
  border-top-left-radius: 3em 1em;
  border-top-right-radius: 2em 4em;
  border-bottom-right-radius: 5em 1em;
  border-bottom-left-radius: 2em 4em;
}

```

20-
Right now, I think your pull request is considerably 
too big (83 files). Can you put all of the parsing tests
(computed, valid, invalid) into a distinct and separate
pull request? I may be able to approve those parsing tests.
And then I will be able to concentrate on your other remaining tests and references.

21-
One way to help reviewers is to put all the tests
and references into an accessible website. Eg
http://www.gtalbot.org/BrowserBugsSection/CSS3Backgrounds/
http://www.gtalbot.org/BrowserBugsSection/CSS3WritingModes/
That way, the reviewer can click and load the tests and references
into a browser to see the files. Otherwise, the reviewer
has to do a lot of clicking, loading, file creating, uploading.
I have started to do this with your tests and references:
http://www.gtalbot.org/BrowserBugsSection/CSS4BordersBoxDecorations/
I have appended "-SZ" to your tests and references.

TalbotG commented:
22-
The references
border-radius-side-shorthands-001-ref.html
and
border-radius-side-shorthands-002-ref.html
are identical. So, best is to remove -002
and reused -001 accordingly.

TalbotG commented:
23-
In
parsing/border-clip-computed.html
at line 20, I see
test_computed_value("border-clip", "10px 1em 10% 1fr 2fr", "10px 20px 10% 1fr 2fr");
this implies that initial, default font size of body text is assumed to be 20px in the test. The problem here is that initial, default font size for unstyled body text in CSS tests are supposed to be 16px. 16px is the default de facto web browser industry font size for unstyled body text.

"The medium font-size computes to 16px."
coming from
http://web-platform-tests.org/writing-tests/assumptions.html

TalbotG commented:
24-
Where (in which file) is declared the test_computed_value() function?

test_computed_value() function is declared in all border-*-computed.html tests in /parsing/ but I can not find test_computed_value() function in any of the .js files.

In 
http://wpt.live/css/support/parsing-testcommon.js
we can see test_valid_value() and test_invalid_value() but not test_computed_value()

TalbotG commented:
25-
In parsing/border-clip-valid.html
the function called in line 13 to line 19 is 

test_computed_value()

when it should be

test_valid_value()

I have not checked other tests.

TalbotG commented:
Sebastian,
I expect to be less available in the next upcoming 4-5 days. I will get back to you on Friday February 9th or Saturday February 10th.

TalbotG commented:
26-

Some (several?) of your parsing tests are _not_ supported by the current spec.

This section
https://drafts.csswg.org/css-borders-4/#index-defined-here
lists all of the property names covered by the spec.  'border-radius-*' do **not** exist (eg.border-radius-bottom , border-radius-top , border-radius-left, etc.: these are not listed in the spec). So, the problem is how (or why) can you be testing, for example, 'border-radius-bottom'
in the filename border-radius-bottom-valid.html ? Same thing with several other tests in /parsing/ . 

There is no problem with testing 'border-bottom-radius' though.

SebastianZ commented:
Thanks again, @TalbotG, for the review and all the detailed feedback! I am still working on fixing all the things you mentioned and hope to finish them tomorrow, Saturday February 10th.

Sebastian

SebastianZ committed:
Fixed corner-shape tests and added references

SebastianZ committed:
Fixed computed, valid, and invalid values tests for CSS Borders 4 properties

SebastianZ committed:
Renamed border-radius-* properties to border-*-radius

SebastianZ committed:
Added assertions to computed value tests for CSS Borders 4

TalbotG commented:
Sebastian,

There is now a lot of work involved in our tests and overall they look much better.

3 quick comments.

a) I do not know, I can not explain now why your edb625f commit failed.

b) parsing/border-radius-inline-end-invalid.html 
was filename-renamed as
parsing/border-inline-end-radius-invalid.html
; it was **only filename-renamed**; the code inside was not updated accordingly.

c) In my opinion, meta assert text should be edited only if a test is complex, involves a particular code scenario. If a test is simple or testing something basic, then there is no need for a test assert text.

SebastianZ committed:
Fixed border-*-radius side shorthand tests

SebastianZ committed:
Removed all parsing tests for CSS Borders 4

SebastianZ commented:
> 16- In border-radius-side-shorthands-001.html I see line 77: `border-radius-right: 1 4em / 5em 1em;` The "1" should be "2em" .

Fixed and reworked the tests.
 
> 17- In all of corner-shape-four-values tests (except 001): if there is no reference file for those tests, then you must append "-manual" to such tests: "a -manual suffix after the main filename but before the extension".

I've now added reference files for all the tests using SVGs to show the expected display.

> 18- In corner-shape-four-values-001 meta name="assert" content="This test checks that 'round round round round' is supported for 'corner-shape'." but, in reality, the test is currently passed by Firefox and Chromium because of a fallback feature: ...

I've removed those tests.

> 19- In tests and/or in references, if you can combine selectors, then this is preferable. Eg In border-radius-side-shorthands-001-ref.html in line 20 to line 46, the code can become:

Done. (Though I've also reworked some of the tests to improve verifiability, so some selectors could not be combined anymore.)

> 20- Right now, I think your pull request is considerably too big (83 files). Can you put all of the parsing tests (computed, valid, invalid) into a distinct and separate pull request? I may be able to approve those parsing tests. And then I will be able to concentrate on your other remaining tests and references.

Fair point. I've split the parsing tests into #44525.

> 21- One way to help reviewers is to put all of your tests and references into an accessible website.

I don't have my own website. So I'll need to check where to upload those, maybe GitHub.io. For now, you'll have to upload them on your site to check them or run a local server, sorry!

> 22- The references border-radius-side-shorthands-001-ref.html and border-radius-side-shorthands-002-ref.html are identical. So, best is to remove -002 and reused -001 accordingly.

Done.

> 23- In parsing/border-clip-computed.html at line 20, I see test_computed_value("border-clip", "10px 1em 10% 1fr 2fr", "10px 20px 10% 1fr 2fr"); this implies that initial, default font size of body text is assumed to be 20px in the test. The problem here is that initial, default font size for unstyled body text in CSS tests is supposed to be 16px. ...

Not sure why I assumed 20px. I think I initially had set `font-size: 20px;` on the element. I changed the expected values to 16px now.

> 24- Where (in which .js file) is declared the test_computed_value() function?
> ...
> > So you need to make appropriate changes in all of your parsing/border-*-computed.html tests. [/Edit]

Fixed.

> 25- In parsing/border-clip-valid.html the function called in line 13 to line 19 is
> 
> test_computed_value()
> 
> when it should be
> 
> test_valid_value()
> 
> I have not checked other tests.

Copy & paste error. Thanks for noticing! Fixed.

> 26-
> 
> Several of your parsing tests are _not_ supported by the current spec.
> 
> This section https://drafts.csswg.org/css-borders-4/#index-defined-here lists each individual property names covered by the spec. 'border-radius-*' do **not** exist (eg.border-radius-bottom , border-radius-top , border-radius-left, etc.: these are not listed in the spec). So, the problem is how (or why) can you be testing, for example, 'border-radius-bottom' in the filename border-radius-bottom-valid.html ? Same thing with several other tests in /parsing/ .

They were initially defined as `border-radius-*` but later got renamed to `border-*-radius` for consistency with the existing `border-*-radius` properties. I've adjusted the tests now to fit to the names defined in the spec.

> a) I do not know, I can not explain now why commit edb625f failed.

I was still in the middle of reworking the tests. It looks like they passed now.

> b) parsing/border-radius-inline-end-invalid.html was filename-renamed as parsing/border-inline-end-radius-invalid.html ; it was **only filename-renamed**; the code inside was not updated accordingly.

Same as above. Please check #44525.

> c) In my opinion, meta assert text should be edited only if a test is complex, involves a particular code scenario. If a test is simple or testing something basic, then there is no need for a test assert text.

I thought I'd better add one to all the tests to let testers know what's expected. Though let me know if I should rather remove them.

Sebastian

TalbotG commented:
> put all of your tests and references into an accessible website. (...)
> you'll have to upload them on your site to check them

I can do that. I will do that. No problem.

SebastianZ committed:
Renamed the corner-shape tests to match the reference files

SebastianZ committed:
Moved all CSS Borders 4 tests to tentative folder

SebastianZ commented:
> I am afraid that you will have to filename-rename all your corner-shape-* tests. When creating a reference file, you must reuse the filename of the associated test.

Dunnit. I also moved all the tests into a `tentative/` folder, now. Hope the changes meet the expectations now.

Sebastian

SebastianZ committed:
Moved CSS Borders 4 tests to css-borders

dbaron commented on css/css-borders/tentative/border-radius-side-shorthands/border-radius-side-shorthands-001.html:
```diff
@@ -0,0 +1,84 @@
+#refenence-shape-2,
+#refenence-shape-3,
+#refenence-shape-4 {
+  border: 10px solid red;
```
I'm a little concerned about the test assertion here.  With antialiased edges, the idea that drawing green thing on top of identically-shaped red thing will cover all of the red isn't really true, since antialiasing is done on the assumption that the coverage areas are independent.

Maybe the answer is that once these features are implemented, browser developers will just add an appropriate `meta name=fuzzy` annotation to all of these tests.

But probably a better answer is -- at least if it's possible -- to make the *reference* do the exact same drawing of a green thing on top of an identically shaped red thing, so that the reference has the same visual artifacts that the test has, and thus the test and reference should really match.

TalbotG replied:
I agree with you on the antialiasing issue. But I believe this can be worked out or adjusted later. Sebastian's tests do not absolutely need to use the overlapping technique for his tests... and if his tests need to, I am sure we can find a work-around for him.

SebastianZ replied:
I just got to know about `meta name=fuzzy` the other day. I might that, though I'd very much need to guess the values. Therefore, it might be better to adjust the tests once implementors start to work on that.

dbaron commented on css/css-borders/tentative/corner-shape/corner-shape-angle-001.html:
```diff
@@ -0,0 +1,22 @@
+<title>CSS Borders and Box Decorations 4 Test: 'corner-shape: angle'</title>
+<link rel="author" title="Sebastian Zartner" href="mailto:sebastianzartner@gmail.com">
+<link rel="help" href="https://drafts.csswg.org/css-borders-4/#corner-shaping">
+<link rel="match" href="../../reference/corner-shape-angle-ref.html">
```
I'm extremely skeptical that you're going to get an exact pixel match between CSS borders and SVG, in the presence of antialiasing.

Maybe the only real answer is that this test will eventually be `fuzzy` once the feature is implemented, as I mentioned above.

(The deeper risk is that a browser developer implements the feature, finds the tests that test the feature by seeing what WPT tests pass once the feature is implemented, and misses these tests because they *don't* start passing... and then the feature later regresses in ways that these tests would have caught if they had been marked as passing when the feature was initially implemented.)

SebastianZ replied:
> (The deeper risk is that a browser developer implements the feature, finds the tests that test the feature by seeing what WPT tests pass once the feature is implemented, and misses these tests because they don't start passing... and then the feature later regresses in ways that these tests would have caught if they had been marked as passing when the feature was initially implemented.)

I assumed that implementors explicitly search for the tests that check the feature they are implementing. And with the name "corner-shape" in the file names it should be clear enough. Though I see your point here.

As noted above, I could add fuzzy matching, though I'd need help to find proper values for `maxDifference` and `totalPixels`.

TalbotG replied:
> though I'd need help to find proper values for maxDifference and totalPixels.

The maxDifference and totalPixels values are and would be indicated in the bottom-left corner of the Compare [screenshots] webpage. An example with an entirely randomly chosen test with failures :

Go to
https://wpt.fyi/results/css/css-backgrounds/background-position-negative-percentage-comparison-002.html
Click one of the Compare links
Look at bottom left corner for those values.

dbaron reviewed (commented):
A few comments on the tests below; I have not reviewed the entire diff, although I'm not sure that I really need to.  A number of the comments apply to multiple tests in (hopefully obvious) ways.

One deeper concern with writing tests like this is that it's *very* hard to check them in the absence of an implementation to verify them.  Hopefully the first implementor (and, of slightly lesser importance, later implementors) of the feature will be aware that these tests exist and thus be able to verify them properly.

At some point I think code review of tests has somewhat limited value, since it's easy to look at them and think they're right, and then later write an implementation following the rules of the spec and discover that they were actually wrong.  So I do bias towards not doing lots of review iteration on tests and just trying to get them landed.

However, as noted, I do have a few comments, but you should feel free to push back if you'd prefer not to make changes.

dbaron commented:
Also, it's worth noting that you [can see](https://wpt.fyi/results/css/css-borders/tentative?label=pr_head&max-count=1&pr=42345) the current success/failure of the modified tests in browsers; this link can be found in the "wpt.fyi" part of the Checks tab above.

TalbotG commented on css/css-backgrounds/border-radius-greater-than-width.html:
> In general, it's probably a good idea to avoid renaming tests unless there's a very good reason to do so. In particular, there are multiple systems that maintain metadata about tests (for example, connections between known failing tests and browser bugs) that would get lost and need to be recreated when tests are renamed.

First, none of Sebastian's Box Decorations tests have been approved and git-incorporated into /css-backgrounds/ or /css-borders/ so far, as far as today. So, what you are describing can not happen.
Second, the base of filename of tests must match the base filename of its associated reference file: that is what the documentation states and that is what a large collection of testers have been doing for over 10 years now.
If a test filename is abc-123.html and its reference is xyz-789-ref.html , then one of the 2 will have to be filename-renamed. That is what I have explained in a comment.

dbaron replied:
The PR as it exists right now (since [c4ecafd](https://github.com/web-platform-tests/wpt/pull/42345/commits/c4ecafded07396b2981d8c8bb09eff026e4cba6c)) is moving existing tests; I was able to quickly find [an external reference](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/TestLists/highdpi;l=313;drc=0e27f4923f27a9e5645403108840b73e606b51f3) to one of the moved files.

TalbotG replied:
The PR exists. The c4ecafd commit exists. But no one approve the related, involved tests so far. Making an external reference in such situation is too soon, not justified and premature. Do you agree?

dbaron replied:
No, I don't agree.  I'm saying that this PR is moving tests that are currently in WPT.

SebastianZ replied:
@dbaron Thank you for noting that there are external references relying on the existing WPTs!
For now, I can revert the change related to them. Though I'd really like to move them where they belong. I assume there's an established process for moving or renaming tests, right?

SebastianZ replied:
I moved them back in https://github.com/web-platform-tests/wpt/pull/42345/commits/a5bf14d68d73a5d8234cc940851ab84be8e3ad41.

TalbotG commented on css/css-borders/tentative/border-radius-side-shorthands/border-radius-side-shorthands-001.html:
```diff
@@ -0,0 +1,84 @@
+<title>CSS Borders and Box Decorations 4 Test: Physical 'border-*-radius' shorthands set the two radii of one side.</title>
+<link rel="author" title="Sebastian Zartner" href="mailto:sebastianzartner@gmail.com">
+<link rel="help" href="https://drafts.csswg.org/css-borders-4/#corner-sizing">
+<link rel="match" href="../../reference/border-radius-side-shorthands-ref.html">
```
> having the references so far away from the tests

That is because he had to put them into a /tentative/ folder. And such folder is bound to be removed once the spec is more reliable, is less "drafty".

TalbotG commented on css/css-borders/tentative/border-radius-side-shorthands/border-radius-side-shorthands-001.html:
```diff
@@ -0,0 +1,84 @@
+  margin: 10px;
+}
+
+#refenence-shape-1,
```
I saw that too but I did not want to annoy Sebastian with a purely textual nit like that.

SebastianZ replied:
No worries. I'll fix that.

TalbotG commented:
@dbaron David,

If you want to review Sebastian's Box Decoration tests, all of them, including those in #44525, go ahead. No problem with me.

SebastianZ commented:
> One deeper concern with writing tests like this is that it's very hard to check them in the absence of an implementation to verify them.

True, my hope is to encourage implementations with those tests.

(I'm aware that I still need to get the spec. in shape to get rid of this nasty "Not Ready For Implementation" banner and be able to request a FPWD.)

> Hopefully the first implementor (and, of slightly lesser importance, later implementors) of the feature will be aware that these tests exist and thus be able to verify them properly.

Isn't it common that the spec. editors create the WPTs and implementors try to match those or refine/fix them if they can't be matched?

Sebastian

SebastianZ commented:
> Also, it's worth noting that you [can see](https://wpt.fyi/results/css/css-borders/tentative?label=pr_head&max-count=1&pr=42345) the current success/failure of the modified tests in browsers; this link can be found in the "wpt.fyi" part of the Checks tab above.

Thank you for the hint! I didn't realize that.

SebastianZ committed:
Fixed spelling mistake for "reference"

TalbotG commented:
Sebastian,

Right now, I believe your tests should be reviewed one last time. I will do that this upcoming weekend (march 23rd and 24th). Unless I receive some kind of objection, complain or suggestion and unless I find something that can be improve in your tests, I will approve PR42345 before march 28th 2024. You have waited a long time before getting a single review, you complied with review comments and the latest version of your tests seem excellent and adequate. There is no apparent or serious reason to delay furthermore this PR.

SebastianZ commented:
Thanks a lot @TalbotG for helping me to get these tests right!
If there's anything left, please let me know! Though, I'm on PTO right now, so my response times might be a bit slow.

Sebastian

TalbotG commented:
Sebastian,

I am in the process of updating my

http://www.gtalbot.org/BrowserBugsSection/CSS4BordersBoxDecorations/

- - - 

27-
In your
reference/corner-shape-angle-bottom-left-001-ref.html
line 6: `<style>`
`.expected {`
`  display: block;`
`  margin-top: 20px;`
`}`
`</style>`

`<p>Test passes if the square has a beveled bottom-left corner and the others are rounded.`

line 15: `<img src="../support/corner-shape-angle-bottom-left.svg" alt="" class="expected">`

The `20px` value is wrong. Vertical margins of `<p>` element is 1em, which is 16px. 

You can reduce amount of code (no class selector, no CSS declaration) and correct this by using the following:

`<p>Test passes if the square has a beveled bottom-left corner and the others are rounded.`

line 8: `<p><img src="../support/corner-shape-angle-bottom-left.svg" alt="Image download support must be enabled">`

which is what I have in

http://www.gtalbot.org/BrowserBugsSection/CSS4BordersBoxDecorations/reference/GT-corner-shape-angle-bottom-left-ref-GT.html

The alt attribute value is also important as it will be useful in case the link to the src attribute value is incorrect or if the /support folder has moved or something else.

I have not checked other references... I am in the process of checking other references..

TalbotG commented:
> In general, it's probably a good idea to avoid renaming tests unless there's a very good reason to do so.

In commit c4ecafd , Sebastian moved some tests
border-radius-greater-than-width.html
subpixel-borders-with-child-border-box-sizing.html
subpixel-borders-with-child.html
and their associated reference files
which were not his; they were authored by Stephen Chenney or  Lin Peng. As I understand this, they were then moved back to their original and correct folder in commit a5bf14d . I assumed and presumed that moving such 3 tests was unintentional and just caused by inexperience with (or newness with) github.

Regarding Sebastian's own tests, he had all the necessary and sufficient reasons to rename them.

TalbotG commented:
Overall, I got confused with "renaming" versus "moving": to me, it is not the same. And tests that are in the master branch (or trunk) and tests that are in a "personal" branch (not merged in master) are not to be confused.

> I'd really like to move them where they belong. I assume there's an established process for moving or renaming tests, right?

If you believe that some other files - not authored by yourself - should be moved into another folder, then you should create another, distinct and separate, PR for such purpose and then notify their original authors. There is no known or official set of rules/protocols for moving tests around but the above is what I would do.

SebastianZ committed:
Removed unnecessary CSS from CSS Borders 4 tests

TalbotG commented:
@SebastianZ ,

Did you check my March 24th comment and item 27? I am asking because I am waiting for you to correct all of the reference/corner-shape-*-ref.html . I want and hope to be able to approve Pull request 42345 for good once all of the reference/corner-shape-*-ref.html become reliable.

I must take a break-vacations from web and internet for a week. I will be back around Friday May 17th.

SebastianZ commented:
@TalbotG Please excuse the long silence on my side! I was bound to other tasks lately. I'll try to fix the remaining issues in the next few days.

Sebastian

SebastianZ committed:
Removed incorrect top margin from

SebastianZ commented:
@TalbotG Ok, I think I incorporated all your last suggestions. I also did a rebase and removed the file incorrect file movement commits completely, so we should be good on that.
I'll review everything again tomorrow but you can already have a look.

Sebastian

SebastianZ committed:
Added fuzziness to border-radius side shorthands and corner-shape tests

SebastianZ commented:
@TalbotG I didn't see any further issues. Please let me know if I missed anything!

@dbaron I did add some fuzziness to the tests now in https://github.com/web-platform-tests/wpt/pull/42345/commits/48b9f4acb5caee1f54b6610cb9b8217c3652bc94 based on some educated guesses looking at the reference renderings, calculating the pixels drawn for anti-aliasing and adding some buffer.
I hope that's fine in the end, though the values may still be adjusted later on.

Sebastian

TalbotG commented:
Sebastian,

commit [b119144](https://github.com/web-platform-tests/wpt/pull/42345/commits/b1191447d79a15f46da9621c2f1d33d803049c62) failed... and, right now, I do not know why.

"
Some checks were not successful
10 cancelled, 1 queued, and 3 successful checks
"

https://github.com/web-platform-tests/wpt/runs/25140115232

The Azure Pipeline checks got queued and never ended...

- - -

At line 8 of  
css/css-borders/reference/corner-shape-angle-bottom-left-001-ref.html
`class="expected"`
can be safely removed... This is a minor oversight without consequences.

- - -

Regarding fuzziness matching in tests and proper values for maxDifference and totalPixels.
 
The proper sequence of events should be: 

1- Browser manufacturers implement new property with its correspondent property values
2- An author like you create tests for such new property with its correspondent property values
3- Then comparison of screen shots of failed tests are available and, in a new and distinct Pull Request, maxDifference and totalPixels values are added to tests that a reviewer estimates adequate. Sometimes the tests (or the reference files) have to be redesigned in order to be more precise. Sometimes, I guess, a test must become a manual test.

In your case, no browser manufacturer has so far implemented any of the property values. So, it is premature to come up with maxDifference and totalPixels values.
`totalPixels=0-3844`
in both of your 
border-radius-side-shorthands-001.html
border-radius-side-shorthands-002.html
seems a lot... but it is too soon to say (or to be definitive about) and I am not a good reviewer regarding fuzziness (caused by antialiased edges).

- - -

I am still looking, searching why 
commit [b119144](https://github.com/web-platform-tests/wpt/pull/42345/commits/b1191447d79a15f46da9621c2f1d33d803049c62) failed...

TalbotG commented:
@SebastianZ Sebastian,

@whimboo Henrik says in [this comment in Issue 46378](https://github.com/web-platform-tests/wpt/issues/46378#issuecomment-2123932260) that we need to close the Pull Request and reopen it in order to trigger the Continuous Integration checks, especially for commit https://github.com/web-platform-tests/wpt/commit/b1191447d79a15f46da9621c2f1d33d803049c62 which obviously has problems (" Some checks were not successful
10 cancelled, 1 queued, and 3 successful checks")

So I am going to do it now... [fingers crossed]

TalbotG commented:
Closing PR now

TalbotG commented:
Reopening PR now

dbaron commented:
Yeah, I agree the pixel counts for those fuzziness values look surprisingly large; I agree that it's better not to include them for now, though it might be nice to leave a comment that says that you know the test will eventually need a meta name=fuzzy, but you just don't know what the required numbers are.

SebastianZ committed:
Reverted addition of fuzziness to border-radius side shorthands and corner-shape tests

Instead of adding <meta name="fuzzy"> tags with guessed values, there's now a todo comment indicating that the tests need to be updated to use fuzzy matching.

SebastianZ commented:
@dbaron I tried my best to estimate the required fuzziness by counting pixels, though I guess my buffer of two times the value was too big. Anyway, I've done now as suggested and replaced the `<meta name="fuzzy">` tags by todo comments.

@TalbotG Excuse my long silence on this again! According to @foolip it sounds like the failing commit doesn't matter because the commits will be squashed when merging them. Though let's see if that works. If not, I can close this PR and create a new one with the changes.

Sebastian

TalbotG commented:
I am ready to approve all this (72 conversations, 59 files, 19 commits, many corrections, 5½ months long). [I am opening a bottle of champagne right now :) ]

From now on, Sebastian, please try to create smaller pull requests.

TalbotG reviewed (approved):
Approving.

TalbotG commented:
@SebastianZ Sebastian,

All of the corner-shape tests results should _eventually_ be listed in a corner-shape folder inside 

https://wpt.fyi/results/css/css-borders/tentative

and, right now, they are **not**.

The corner-shape tests are in:

http://wpt.live/css/css-borders/tentative/corner-shape/

and their associated reference files are in:

http://wpt.live/css/css-borders/reference/

So far, after some random visual check on my part, everything seems correct (including the removal of `margin-top: 20px` in the reference files), good and as expected, except the listing of the corner-shape tests results.

SebastianZ commented:
> I am ready to approve all this (72 conversations, 59 files, 19 commits, many corrections, 5½ months long).

Thank you again for all your efforts and help, @TalbotG! I learned a lot from all this!

> [I am opening a bottle of champagne right now :) ]

Cheers!😄🍾 

> From now on, Sebastian, please try to create smaller pull requests.

Promise!

Sebastian

SebastianZ commented:
> All of the corner-shape tests results should _eventually_ be listed in a corner-shape folder inside
> 
> https://wpt.fyi/results/css/css-borders/tentative

I'll observe that.

Sebastian

TalbotG commented:
All of the corner-shape tests results are now listed in
https://wpt.fyi/results/css/css-borders/tentative/corner-shape

TalbotG commented:
In 
corner-shape/corner-shape-round-top-left-001.html
http://wpt.live/css/css-borders/tentative/corner-shape/corner-shape-round-top-left-001.html
we see "other**e**s" 

https://wpt.fyi/results/css/css-borders/tentative/corner-shape/corner-shape-round-top-left-001.html

When you can, create a Pull request to remove that "e" and, at the same time, remove
`class="expected"` from
http://wpt.live/css/css-borders/reference/corner-shape-angle-bottom-left-001-ref.html

and then add me as a reviewer to such Pull request.

SebastianZ commented:
@TalbotG You obviously have eagle eyes.😄
Here's the fix: https://github.com/web-platform-tests/wpt/pull/46574.

Sebastian
